rosbag2_transport/test/rosbag2_transport/test_keyboard_controls.cpp

```diff
@@ -0,0 +1,165 @@
+
+  auto primitive_message2 = get_messages_strings()[0];
+  primitive_message2->string_value = "Hello World 2";
+
+  auto message_time_difference = rclcpp::Duration(1, 0);
+  auto topics_and_types =
+    std::vector<rosbag2_storage::TopicMetadata>{{"topic1", "test_msgs/Strings", "", ""}};
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> messages =
+  {serialize_test_message("topic1", 0, primitive_message1),
+    serialize_test_message("topic1", 0, primitive_message2)};
+
+  messages[0]->time_stamp = 100;
+  messages[1]->time_stamp = messages[0]->time_stamp + message_time_difference.nanoseconds();
+
+  play_options_.play_next_key = KeyboardHandler::KeyCode::UNKNOWN;
+
+  auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+  prepared_mock_reader->prepare(messages, topics_and_types);
+  auto reader = std::make_unique<rosbag2_cpp::Reader>(std::move(prepared_mock_reader));
+
+  EXPECT_THROW(
+    std::make_shared<Player>(
+      std::move(reader), storage_options_, play_options_),
+    std::invalid_argument
+  );
+}
+
+TEST_F(RosBag2PlayTestFixture, test_keyboard_controls)
+{
+  auto primitive_message1 = get_messages_strings()[0];
+  primitive_message1->string_value = "Hello World 1";
+
+  auto primitive_message2 = get_messages_strings()[0];
+  primitive_message2->string_value = "Hello World 2";
+
+  auto primitive_message3 = get_messages_strings()[0];
+  primitive_message3->string_value = "Hello World 3";
+
+  auto message_time_difference = rclcpp::Duration(1, 0);
+  auto topics_and_types =
+    std::vector<rosbag2_storage::TopicMetadata>{{"topic1", "test_msgs/Strings", "", ""}};
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> messages =
+  {serialize_test_message("topic1", 0, primitive_message1),
+    serialize_test_message("topic1", 0, primitive_message2),
+    serialize_test_message("topic1", 0, primitive_message3)};
+
+  messages[0]->time_stamp = 100;
+  messages[1]->time_stamp = messages[0]->time_stamp + message_time_difference.nanoseconds();
+  messages[2]->time_stamp = messages[1]->time_stamp + message_time_difference.nanoseconds();
+
+  auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+  prepared_mock_reader->prepare(messages, topics_and_types);
+  auto reader = std::make_unique<rosbag2_cpp::Reader>(std::move(prepared_mock_reader));
+  auto keyboard_handler = new MockKeyboardHandler(
+    play_options_.pause_resume_toggle_key,
+    play_options_.play_next_key);
+
+  // play bag in thread
+  auto player = std::make_shared<CountPlayer>(
+    std::move(reader), std::shared_ptr<KeyboardHandler>(keyboard_handler),
+    storage_options_, play_options_);
+  std::thread player_thread = std::thread([player]() {player->play();});
+
+  EXPECT_THAT(player->is_paused(), false);
+  keyboard_handler->play_next();
+  EXPECT_THAT(player->is_paused(), false);
+  keyboard_handler->pause_resume();
+  EXPECT_THAT(player->is_paused(), true);
+  keyboard_handler->play_next();
```

Comment:
@lihui815 Could you please redesign `MockKeyboardHandler` such a way that it will have only one extra method 
i.e. 
```cpp
void simulate_key_press(KeyboardHandler::KeyCode key_code, 
                        KeyboardHandler::KeyModifiers key_modifiers = KeyboardHandler::KeyModifiers::NONE);
```
instead of two others `keyboard_handler->play_next();` and `keyboard_handler->pause_resume();`

In tests you can use `play_options_.pause_resume_toggle_key` and `play_options_.play_next_key` as arguments for `keyboard_handler->simulate_key_press(key_code)`

It will be much cleaner to what is going on behind the scene when reading dedicated unit test.